api/document/serializers/content.py

```diff
@@ -102,59 +102,91 @@ def to_internal_value(self, data: PrimitiveDict) -> PrimitiveDict:
         return serializer.to_internal_value(data)
 
 
-class BaseAnnotationSerializer(serializers.Serializer):
-    # This determines whether the annotation is a leaf of our
-    # content tree.
-    IS_LEAF = False
+class NestableAnnotationField(serializers.Field):
+    def get_attribute(self,
+                      instance: NestableAnnotation) -> NestableAnnotation:
+        return instance
 
-    content_type = serializers.SerializerMethodField()
-    inlines = serializers.SerializerMethodField()
-    text = serializers.SerializerMethodField()
 
-    @property
-    def CONTENT_TYPE(self):     # noqa; this is a constant
-        raise NotImplementedError('Unknown annotation type')
+class InlinesField(NestableAnnotationField):
+    def __init__(self, is_leaf_node: bool):
+        super().__init__()
+        self.is_leaf_node = is_leaf_node
+
+    def to_representation(self,
+                          instance: NestableAnnotation) -> List[PrimitiveDict]:
+        if self.is_leaf_node:
+            return []
+        return NestedAnnotationSerializer(
+            instance.children, context=self.context, many=True).data
+
+    def to_internal_value(self,
+                          data: List[PrimitiveDict]) -> List[PrimitiveDict]:
+        if not self.is_leaf_node:
+            serializer = NestedAnnotationSerializer()
+            return [
+                serializer.to_internal_value(item) for item in data
+            ]
+        elif data:
+            raise ValidationError('leaf nodes cannot contain nested content')
+        return []
+
+
+class TextField(NestableAnnotationField):
+    def __init__(self, is_leaf_node: bool):
+        super().__init__()
+        self.is_leaf_node = is_leaf_node
 
     @property
     def doc_node_text(self):
         return self.context['cursor'].text
 
+    def to_representation(self, instance: NestableAnnotation) -> str:
+        return self.doc_node_text[instance.start:instance.end]
+
+    def to_internal_value(self, data: str) -> str:
+        if self.is_leaf_node:
+            serializer = serializers.CharField(trim_whitespace=False)
+            return serializer.to_internal_value(data)
```

Comment:
I think that we could potentially just have this field inherit from `CharField`, but there's so much idiosyncratic logic in here that it was easier for me to just instantiate one during this method instead... I could switch things up, though.

Reply:
Good point! I'm also torn; maybe we keep it like this until/if we find ourselves wanting to set more validation properties (maybe some annotations can only be x chars long)?